graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -0,0 +1,94 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file cuda/index_select_impl.cu
+ * @brief Index select operator implementation on CUDA.
+ */
+#include <torch/script.h>
+
+#include <numeric>
+
+#include "../index_select.h"
+#include "./macro.h"
+
+namespace graphbolt {
+namespace ops {
+
+template <typename DType, typename IdType>
+__global__ void IndexSelectMultiKernel(
+    const int64_t input_len, const DType* const input, const int64_t feat_size,
+    const IdType* const index, const int64_t output_len, DType* const output) {
+  int64_t out_row_index = blockIdx.x * blockDim.y + threadIdx.y;
+
+  const int64_t stride = blockDim.y * gridDim.x;
+
+  while (out_row_index < output_len) {
+    int64_t column = threadIdx.x;
+    const int64_t in_row = index[out_row_index];
+    assert(in_row >= 0 && in_row < input_len);
+    while (column < feat_size) {
+      output[out_row_index * feat_size + column] =
+          input[in_row * feat_size + column];
+      column += blockDim.x;
+    }
+    out_row_index += stride;
+  }
+}
+
+template <c10::DeviceType XPU, typename DType, typename IdType>
+torch::Tensor UVAIndexSelectImpl(torch::Tensor input, torch::Tensor index) {
```

Comment:
We can also dispatch data types directly in this function so that template can be ignored, and the below declaration can also be removed. Which one do you think is better?

Reply:
Yeah. I only expose the dispatching of device type of this interface.